Server Error: make sure to pass on valid context

flootr opened the pull request:
• Rename `serverRenderError` to `renderServerError` and move it to `server/pages`
• Make sure to pass on a valid (react) context object to render the 500 component

The latter is likely a fix for the `TypeError: Cannot read property 'style.css' of undefined` which you see sometimes on local calypso builds.

**What (might have) caused the error?**

If we render certain documents on the server (such as the default one, a 404 template or the 500) we expect a valid context object (= _props_ in React speech). It seems like we missed to provide one for rendering the 500 layout.

I think it's still an open question why the 500 got rendered but this PR at least should fix the render error.

Edit: Line 502 to 503 is the bug related change: https://github.com/Automattic/wp-calypso/pull/26655/files#diff-b2d6784517d1151d7e75f0287d0c6478R502

### testing

Go through these steps both on `master` and this branch locally:

1. Modify the function `render404` in `server/pages/index.js` to look like the following:

```js
// server/pages/index.js:490
function render404( request, response, next ) {
	const ctx = getDefaultContext( request );
	return next( new Error( 'some error' ) ); // <-- produce a server error
	response.status( 404 ).send( renderJsx( '404', ctx ) );
}
```

2. Run Calypso with `npm start`
3. After it built, go to `calypso.localhost:3000/blah`
4. Make sure 500 renders correctly (shouldn't on `master`)

For 1. you can likely also use any other method to produce a server error, that's only how I did it.

flootr committed:
Server Error: make sure to pass on valid context

Also, move method to render the server error to server/pages/index.js to have it in the same file as e.g. render404.

ockham commented:
I'll need a bit to mull over this. I know that `serverRenderError` [was changed](https://github.com/Automattic/wp-calypso/commit/faf2acdd020f2c1d1d96f32a0416ffc13d03de6d) at some point that I thought introduced some risks, but it's been a while.

My concerns are mostly as follows: Since the 500 is an emergency hatch scenario, I think we want as few prerequisites to be fulfilled -- `getDefaultContext()` is complex enough itself to cause errors, which AFAICS would then cause the 500 to fail.

I'll give this some more thought and report back.

flootr commented:
@ockham I agree. We somehow need to provide at least `urls` from default context. We might be able to reduce it to just that and minimize the failure surface.

> [...] cause the 500 to fail.

This is what happens currently and this PR is trying to solve. We need to find a way to render 500 without adding new points of failure. Curious what you come up with 👍

ockham commented on server/pages/index.js:
```diff
@@ -492,6 +492,21 @@ function render404( request, response ) {
 }
 
+function renderServerError( err, req, res, next ) {
+	if ( err ) {
```
I know this was simply moved over from [`server/render`](https://github.com/Automattic/wp-calypso/pull/26655/files#diff-d92d930a21711a960b300aeb231d9056L193), but the check doesn't seem to make a lot of sense -- it essentially verifies there's at least one arg, but we're referencing all the other args regardless (most notably, `next()` would fail quite badly if it wasn't present). I'd opt for removing it.

flootr committed:
server: improve ctx for 'serverRenderError' fn

- extract a 'getFaviconUrl' fn to for reuse
- remove 'err' check for 'serverRenderError'

ockham commented on server/pages/index.js:
```diff
@@ -492,6 +492,21 @@ function render404( request, response ) {
+		}
+		req.error = err;
+		res.status( err.status || 500 );
+		const ctx = getDefaultContext( req );
```
Ohh, that's a very good point! Yeah, let's do that -- I don't think there's more to it!

flootr committed:
server: derive favicon url from config

flootr commented:
@ockham I think this is worth another look

ockham commented on config/_shared.json:
```diff
@@ -1,6 +1,7 @@
 {
 	"env": "shared",
 	"env_id": "shared",
+	"favicon_url": "//s1.wp.com/i/favicon.ico",
```
Shouldn't we also add an explicit `favicon_url` to `config/production.json`?

flootr replied:
Not sure. Previously it was the fallback/default as well. I don't mind adding it.

ockham replied:
Yeah, let's please add it. Personally, when I find a config setting in all the other ones, I'd go looking for the production setting in `production.json`.

flootr replied:
done

ockham commented:
I've just followed testing instructions, but I'm not getting the 500 template rendered, but just the error and call stack:

```
Error: some error
    at render404 (~/src/wp-calypso/build/webpack:/server/pages/index.js:489:15)
    at Layer.handle [as handle_request] (~/src/wp-calypso/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (~/src/wp-calypso/node_modules/express/lib/router/index.js:317:13)
    at ~/src/wp-calypso/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (~/src/wp-calypso/node_modules/express/lib/router/index.js:335:12)
    at next (~/src/wp-calypso/node_modules/express/lib/router/index.js:275:10)
    at Layer.handle [as handle_request] (~/src/wp-calypso/node_modules/express/lib/router/layer.js:91:12)
    at trim_prefix (~/src/wp-calypso/node_modules/express/lib/router/index.js:317:13)
    at ~/src/wp-calypso/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (~/src/wp-calypso/node_modules/express/lib/router/index.js:335:12)
    at next (~/src/wp-calypso/node_modules/express/lib/router/index.js:275:10)
    at Layer.handle [as handle_request] (~/src/wp-calypso/node_modules/express/lib/router/layer.js:91:12)
    at trim_prefix (~/src/wp-calypso/node_modules/express/lib/router/index.js:317:13)
    at ~/src/wp-calypso/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (~/src/wp-calypso/node_modules/express/lib/router/index.js:335:12)
    at next (~/src/wp-calypso/node_modules/express/lib/router/index.js:275:10)``

flootr committed:
server(express): make sure to pass all four args to error handler

ockham commented on server/pages/index.js:
```diff
@@ -488,7 +488,9 @@ function render404( request, response ) {
 
-function renderServerError( err, req, res ) {
+/* eslint-disable no-unused-vars */
+function renderServerError( err, req, res, next ) {
```
Right, it's the signature that matters with these error handlers (previouly [broken](https://github.com/Automattic/wp-calypso/pull/26655#issuecomment-414378404)). Good catch!

Let's maybe add a comment to the code so future generations of developers :smile: touching this code will know _why_ there's an ununsed arg. Also, let's change the `eslint-disable`/`eslint-enable` pair to either `eslint-disable-line` (on the same line) or `eslint-disable-next-line` (on the previous line).

flootr replied:
I tried to use `eslint-disable-next-line` but prettier comes in the way and shifts it to the next line. I'll add a comment 👍

ockham reviewed (approved):
Two more requests (and a rebase).

Feel free to merge after addressing :+1:

flootr committed:
config: add 'favicon_url' for prod
